src/vmm/src/device_manager/mmio.rs

```diff
@@ -156,38 +156,43 @@ impl MMIODeviceManager {
         }
     }
 
-    /// Allocates resources for a new device to be added.
-    fn allocate_new_slot(&mut self, irq_count: u32) -> Result<MMIODeviceInfo> {
-        let irqs = self.irqs.get(irq_count)?;
-        let slot = MMIODeviceInfo {
-            addr: self.next_avail_mmio,
-            len: MMIO_LEN,
-            irqs,
-        };
-        self.next_avail_mmio += MMIO_LEN;
-        Ok(slot)
+    // Allocates an irq and registers it with KVM for a given VM, event.
+    fn allocate_irq(&mut self, vm: &VmFd, evt: &EventFd) -> Result<u32> {
```

Comment:
IMO this function should do just what the name says (i.e allocate the irq count that the device needs and let the device choose how many interrupts will use from the ones allocated and how it will use them) so what I propose here is to transform the current `allocate_irq` function you propose into:
```
fn allocate_irqs(&mut self, irq_count: u32) -> Result<Vec<u32>> {
self.irqs.get(irq_count)
}
``` 
and the `vm.register_irqfd` logic is done separately by each device (like it used to be), WDYT?